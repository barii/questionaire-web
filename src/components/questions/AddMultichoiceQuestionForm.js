import React from 'react';
import { connect } from 'react-redux';
import { addMultichoiceQuestion } from '../../actions/multichoiceQuestion';
import { Form, Button, FormControl, FormGroup, Col, ControlLabel } from 'react-bootstrap';
import { v4 as uuid } from 'uuid';

class AddMultichoiceQuestionForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      text: props.question ? props.question.text : '',
      description: props.question ? props.question.description : '',
      choices: props.question ? props.question.choices : [{_id: uuid(), value:''}],
      error: ''
    };
  };
  onTextChange = (e) => {
    const text = e.target.value;
    this.setState(() => ({ text }));
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onChoiceChange = (e, idx) => {
    let choices = this.state.choices;
    choices[idx].value = e.target.value;

    if (choices[choices.length-1].value!=='') {
      choices = [
        ...choices,
        {_id: uuid(), value:""}
      ];
    }

    this.setState(() => ({ choices }));

  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.text) {
      this.setState(() => ({ error: 'Please provide text.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.dispatch(addMultichoiceQuestion({
        description: this.state.description,
        text: this.state.text,
        choices: this.state.choices
      }));
    }
//    this.props.history.push('/');
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <Form onSubmit={this.onSubmit} horizontal> 
          <FormGroup controlId="formInlineName">
            <Col componentClass={ControlLabel} sm={2}>
              Text
            </Col>
            <Col sm={5}>
              <FormControl
                type="text"
                placeholder="Text"
                autoFocus
                value={this.state.text}
                onChange={this.onTextChange}
              />
            </Col>
          </FormGroup>
          <FormGroup controlId="formInlineName">
            <Col componentClass={ControlLabel} sm={2}>
              Description
            </Col>
            <Col sm={5}>
              <FormControl
                type="textarea"
                placeholder="Add a note for your question (optional)"
                value={this.state.description}
                onChange={this.onDescriptionChange}
              />
            </Col>
          </FormGroup>
          <FormGroup controlId="formInlineName">
            <Col  componentClass={ControlLabel} sm={2}>
              Choices
            </Col>
            <Col sm={5}>
            {this.state.choices.map((choice, i) => (
                <FormControl
                  type="text"
                  value={choice.value}
                  key={choice._id}
                  onChange={(e) => this.onChoiceChange(e, i)}
                />
            ))}
            </Col>
          </FormGroup>
          <Button bsStyle="success"  type="submit">Add</Button>
        </Form>
      </div>
    );
  }
}

export default connect()(AddMultichoiceQuestionForm);