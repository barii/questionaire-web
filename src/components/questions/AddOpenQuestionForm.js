import React from 'react';
import { connect } from 'react-redux';
import TextFieldGroup from '../common/TextFieldGroup';
import { Form } from 'react-bootstrap';

class AddOpenQuestionForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      //text: props.question ? props.question.text : '',
      //description: props.question ? props.question.description : '',
      errors: {}
    };

  }

  onChange = (e) => {
    this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.props.question.text) {
      this.setState(() => ({ error: 'Please provide text.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.addQuestion({
        description: this.props.question.description,
        text: this.props.question.text
      });
    }
  };

  render() {
    const { errors } = this.state;

    return (
      <div>
        <Form onSubmit={this.onSubmit} horizontal >
          <TextFieldGroup
            placeholder="Text"
            name="text"
            value={this.props.question.text}
            onChange={this.onChange}
            error={errors.text}
          />
          <TextFieldGroup
            placeholder="Description"
            name="description"
            value={this.props.question.description}
            onChange={this.onChange}
            error={errors.description}
          />

          <input type="submit" bsStyle="success"></input>
        </Form>
      </div>
    )
  }
}

const mapStateToProps = state => ({
  errors: state.errors
});

export default connect(mapStateToProps)(AddOpenQuestionForm);