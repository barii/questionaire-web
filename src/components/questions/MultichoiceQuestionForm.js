import React from 'react';
import TextFieldGroup from '../common/TextFieldGroup';
import ListFieldGroup from '../common/ListFieldGroup';
import { Form, Button, Card, CardHeader, CardBody } from 'reactstrap';


export default function MultichoiceQuestionForm({question, errors, onChange, onSubmit, onAdd}) {
  return (
    <div>
    <Card>
      <CardHeader>Header</CardHeader>
      <CardBody>
        <Form onSubmit={onSubmit} > 
          <TextFieldGroup
            label="Text"
            name="text"
            value={question.text}
            onChange={onChange}
            error={errors.text}
          />
          <TextFieldGroup
            label="Description"
            name="description"
            value={question.description}
            onChange={onChange}
            error={errors.description}
          />
          <ListFieldGroup
            label="Choices"
            name="choices"
            options={question.choices}
            onChange={onChange}
            error={errors.choices}
          />
          <div>
            <Button id="btnAdd" color="primary" onClick={onAdd}>Add</Button>
          </div>
          <div>
            <Button type="submit" color="success">Save</Button>
          </div>
        </Form>
      </CardBody>
    </Card>
  </div>
  );
}
