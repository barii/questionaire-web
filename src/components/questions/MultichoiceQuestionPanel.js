import React from 'react';
import { Button, Card, CardHeader, CardBody } from 'reactstrap';

export default function MultichoiceQuestionPanel({question, onEdit}) {

  return (
    <div>
      <Card>
        <CardHeader>Header</CardHeader>
        <CardBody>
          <div className="form-group">
            <select className="form-control form-control-lg">
              {question.choices.map((choice, i) => (
                <option id={choice.id} key={choice.id}>{choice.value}</option>
              ))}
            </select>
          </div>

          <Button type="submit" color="success" onClick={onEdit}>Edit</Button>
        </CardBody>
      </Card>
    </div>
  )
}


