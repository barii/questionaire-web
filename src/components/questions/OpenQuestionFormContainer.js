import React from 'react';
import uuid from 'uuid';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import OpenQuestionForm from './OpenQuestionForm';

export class OpenQuestionFormContainer extends React.Component {

  constructor(props) {
    super(props);

    this.state = {
      errors: {},
      question: this.props.question
    }
  }

  onChange = (e) => {
    const target = e.target
    this.setState((state) => {
      return {
        question: {
          ...state.question,
          [target.name]:  target.value
        }
      }
    })
  }

  onSubmit = (e) => {
    e.preventDefault();
    let errors = {};

    if (!this.state.question.text) {
      errors.text = 'Please provide text.';
    }

    this.setState({ errors });
    if(Object.keys(errors).length === 0) {
      this.props.onSubmit(this.state.question);
    }
  };

  render() {
      return <OpenQuestionForm 
        question={this.state.question} 
        errors={this.state.errors}
        onChange={this.onChange} 
        onSubmit={this.onSubmit}
      />
  }
}

OpenQuestionFormContainer.defaultProps = {
  question: {
    _id:uuid(),
    type:"open",
    text:'',
    description:''
  }
};

OpenQuestionFormContainer.propTypes = {
  question: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
  }),
  onSubmit: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  errors: state.errors
});

export default connect(mapStateToProps)(OpenQuestionFormContainer);