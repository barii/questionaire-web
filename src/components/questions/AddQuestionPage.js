import React from 'react';
import { connect } from 'react-redux';
import AddOpenQuestionForm from './AddOpenQuestionForm';
import AddMultichoiceQuestionForm from './AddMultichoiceQuestionForm';
import QuestionList from './QuestionList';
import { Jumbotron, PageHeader, FormGroup, FormControl } from 'react-bootstrap';

class AddQuestionPage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      questionType: 'question'
    };
  }
  handleChange = (event) => {
    const questionType = event.target.value;
    this.setState(() => ({questionType}));
  };

  render() {
    let selects="";
    if (this.state.questionType==='question') {
      selects = (<AddOpenQuestionForm />);
    } else {
      selects = (<AddMultichoiceQuestionForm />);
    }
    
    return (
      <div>
      <QuestionList />  
      <Jumbotron>
          <PageHeader>Add Question</PageHeader>
          <FormGroup controlId="formControlsSelectMultiple">
            <FormControl componentClass="select" value={this.state.value} onChange={this.handleChange}>
              <option value="question">Question</option>
              <option value="multichoice-question">Multichoice</option>
            </FormControl>
          </FormGroup>  
          {selects}
        </Jumbotron>
        
      </div>
    )
  };
}

export default connect()(AddQuestionPage);
