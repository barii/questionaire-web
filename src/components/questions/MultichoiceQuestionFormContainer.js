import React from 'react';
import { v4 as uuid } from 'uuid';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import MultichoiceQuestionForm from './MultichoiceQuestionForm';

export class MultichoiceQuestionFormContainer extends React.Component {

  constructor(props) {
    super(props);
    
    this.state = {
      errors: {},
      question: this.props.question
    }

  };

  onChange = (e, idx) => {
    const {name, value} = e.target;
    if(idx===undefined) {
      this.setState(prev => ({
        question: {
          ...prev.question,
          [name]:  value
        }
      }))
    } else {
      this.setState(prev => ({
        question: {
          ...prev.question,
          [name]: prev.question[name].map(f => (f._id === idx) ? {...f, 'value': value} : f)
        }
      }))
    } 
  }

  onAdd = () => {
    this.setState(prev => ({
      question: {
        ...prev.question,
        choices: [
          ...prev.question.choices,
          {_id: uuid(), value:""}
        ]
      }
    }))
  }

  isempty = (e) => {
    return !e.value;
  }

  onSubmit = (e) => {
    e.preventDefault();
    let errors = {};

    if (!this.state.question.text) {
      errors.text = 'Please provide text.';
    }

    if (this.state.question.choices.some(this.isempty)) {
      errors.choices = {};//'Please provide value for all choices.';
      this.state.question.choices.filter(this.isempty).forEach(choice => 
        errors.choices[choice._id] = 'Please provide value for all choices'
      )
    }

    this.setState(() => ({ errors }));

    if(Object.keys(errors).length === 0) {
      this.props.onSubmit(this.state.question);
    }
  };


  

  render() {
    return <MultichoiceQuestionForm 
      question={this.state.question} 
      errors={this.state.errors}
      onChange={this.onChange} 
      onSubmit={this.onSubmit}
      onAdd={this.onAdd}
    />
  }
}


MultichoiceQuestionFormContainer.defaultProps = {
  question: {
    _id: uuid(),
    type:"multichoice",
    text:'',
    description:'',
    choices: [
      {
        _id: uuid(), 
        value:""
      },
      {
        _id: uuid(), 
        value:""
      }
    ]
  }
};

MultichoiceQuestionFormContainer.propTypes = {
  question: PropTypes.shape({
    _id: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    choices: PropTypes.arrayOf(
      PropTypes.shape({
        _id: PropTypes.string.isRequired,
        value: PropTypes.string.isRequired
      })
    )
  }),
  onSubmit: PropTypes.func.isRequired
}

const mapStateToProps = state => ({
  errors: state.errors
});

export default connect(mapStateToProps)(MultichoiceQuestionFormContainer);