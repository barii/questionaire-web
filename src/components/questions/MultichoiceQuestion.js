import React from 'react';
import { connect } from 'react-redux';
import MultichoiceQuestionFormContainer from './MultichoiceQuestionFormContainer';
import MultichoiceQuestionPanel from './MultichoiceQuestionPanel';

class MultichoiceQuestion extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      editing: false,
    };

  }

  // componentDidUpdate() {
  //   if (this.state.question != this.props.question) {
  //     this.setState({ question: this.props.question });
  //   }
  // }

  onEdit = (e) => {
    e.preventDefault();

    this.setState({ editing: true });
  }

  onSubmit = (question) => {
    this.props.onSubmit(question);
    this.setState({ editing: false });
  };

  render() {
    if (this.state.editing) {
      return <MultichoiceQuestionFormContainer
        question={this.props.question} 
        onSubmit={this.onSubmit}
      />
    } else {
      return <MultichoiceQuestionPanel 
        question={this.props.question} 
        onEdit={this.onEdit}
      />
    }
  }
}

const mapStateToProps = state => ({
  errors: state.errors
});

export default connect(mapStateToProps)(MultichoiceQuestion);