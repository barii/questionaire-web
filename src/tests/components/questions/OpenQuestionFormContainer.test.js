import React from 'react';
import { shallow } from 'enzyme';
import { checkProps } from '../../testUtils'
import { OpenQuestionFormContainer } from '../../../components/questions/OpenQuestionFormContainer';


const question = {
  type: "OpenQuestion",
  id: '1',
  text: "Text",
  description: ""
}
const questionWithEmptyText = {
  type: "OpenQuestion",
  id: '1',
  text: "",
  description: ""
}

const onSubmit = jest.fn();
const setup = (initialState={}) => {
  return shallow(<OpenQuestionFormContainer {...initialState} onSubmit={onSubmit} />);
} 

it('should not throw warning with expected props', () => {
  checkProps(OpenQuestionFormContainer, {question, onSubmit})
});

it('should render OpenQuestionFormContainer correctly (SNAPSHOT)', () => {
  const wrapper = setup({question});
  expect(wrapper).toMatchSnapshot();
});

it('should have a question, but no error in state after creation', () => {
  const wrapper = setup({question});
  expect(wrapper.state('question')).toEqual(question);
  expect(wrapper.state('errors')).toEqual({});
});

it('should update state.name correctly after content of textbox is changed', () => {
  const wrapper = setup({question});
  const newText = 'new Text';
  wrapper.instance().onChange({ target: {name: 'text', value: newText }} );
  expect(wrapper.state('question')).toEqual({...question, text: newText});
});

it('should update state.desctiption correctly after content of textbox is changed', () => {
  const wrapper = setup({question});
  const newDescription = 'new description';
  wrapper.instance().onChange({ target: {name: 'description', value: newDescription }} );
  expect(wrapper.state('question')).toEqual({...question, description: newDescription});
});

it('should throw an error if submitted with emoty text', () => {
  const wrapper = setup({question: questionWithEmptyText});
  wrapper.instance().onSubmit({ preventDefault() {} } );

  expect(wrapper.state('errors')).not.toEqual({});
  expect(onSubmit).not.toHaveBeenCalled();
});

it('should be submitted with correct state', () => {
  const wrapper = setup({question});

  wrapper.instance().onSubmit({ preventDefault() {} } );

  expect(wrapper.state('errors')).toEqual({});
  expect(onSubmit).toHaveBeenCalled();
});

