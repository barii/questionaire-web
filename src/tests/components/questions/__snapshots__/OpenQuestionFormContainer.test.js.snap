// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render OpenQuestionFormContainer correctly (SNAPSHOT) 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OpenQuestionFormContainer
    onSubmit={[Function]}
    question={
        Object {
            "description": "",
            "id": "1",
            "text": "Text",
            "type": "OpenQuestion",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "errors": Object {},
      "onChange": [Function],
      "onSubmit": [Function],
      "question": Object {
        "description": "",
        "id": "1",
        "text": "Text",
        "type": "OpenQuestion",
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errors": Object {},
        "onChange": [Function],
        "onSubmit": [Function],
        "question": Object {
          "description": "",
          "id": "1",
          "text": "Text",
          "type": "OpenQuestion",
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
